02-peer-protocol.md

```diff
@@ -1508,23 +1509,68 @@ Closing happens in two stages:
 2. once all HTLCs are resolved, the final channel close negotiation begins.
 
         +-------+                              +-------+
-        |       |--(1)-----  shutdown  ------->|       |
-        |       |<-(2)-----  shutdown  --------|       |
+        |       | shutdown(scriptA1)           |       |
```

Comment:
Nice diagram! And I think it may be more helpful if we could break it down case by case, also addresses the [comment here](https://github.com/lightning/bolts/pull/1096#discussion_r1483925909), to make it clear that,
1. a pair of `closing_complete` and `closing_sig` completes the flow, and,
2. whoever sends the `closing_complete` pays the fees.

In a hindsight it probably makes more sense to name `closing_complete` `closing_params` and `closing_sig` `closing_complete`, well, but the code is written so nvm.

Reply:
During normal flow, once the shutdown scripts are exchanged via `shutdown` messages, Alice sends a `closing_complete` to specify the fees she is willing to pay, and Bob replies with a `closing_sig` to consent.

```
 +-------+                              +-------+
 |       | shutdown(scriptA1)           |       |
 |       |----------------------------->|       |
 |       |           shutdown(scriptB1) |       |
 |       |<-----------------------------|       |
 |       |                              |       |
 |       | <complete all pending HTLCs> |       |
 |   A   |             ....             |   B   |
 |       |                              |       |
 |       | closing_complete             |       |
 |       |----------------------------->|       |
 |       |                  closing_sig |       |
 |       |<-----------------------------|       |
 +-------+                              +-------+
 ```

If Bob disagrees with closing fees, he can send a `closing_complete` to specify the fees he is willing to pay. Alice should reply with a `closing_sig` if she agrees.

 ```
 +-------+                              +-------+
 |       | shutdown(scriptA1)           |       |
 |       |----------------------------->|       |
 |       |           shutdown(scriptB1) |       |
 |       |<-----------------------------|       |
 |       |                              |       |
 |       | <complete all pending HTLCs> |       |
 |   A   |             ....             |   B   |
 |       |                              |       |
 |       | closing_complete             |       |
 |       |----------------------------->|       |
 |       |             closing_complete |       |
 |       |<-----------------------------|       |
 |       | closing_sig                  |       |
 |       |----------------------------->|       |
 +-------+                              +-------+
 ```

It's allowed to specify shutdown scripts multiple times.

 ```
 +-------+                              +-------+
 |       | shutdown(scriptA1)           |       |
 |       |----------------------------->|       |
 |       |           shutdown(scriptB1) |       |
 |       |<-----------------------------|       |
 |       |                              |       |
 |       | <complete all pending HTLCs> |       |
 |   A   |             ....             |   B   |
 |       |                              |       |
 |       |   <A updates their script>   |       |
 |       |                              |       |
 |       | shutdown(scriptA2)           |       |
 |       |----------------------------->|       |
 |       |           shutdown(scriptB1) |       | (*) Bob doesn't update his script
 |       |<-----------------------------|       |
 |       | closing_complete             |       |
 |       |----------------------------->|       |
 |       |                  closing_sig |       |
 |       |<-----------------------------|       |
 +-------+                              +-------+
 ```

It's possible to encounter a race condition when updating the shutdown script. This is handled by always sending an updated `closing_complete` with the latest shutdown script used.

 ```
 +-------+                              +-------+
 |       | shutdown(scriptA1)           |       |
 |       |----------------------------->|       |
 |       |           shutdown(scriptB1) |       |
 |       |<-----------------------------|       |
 |       |                              |       |
 |       | <complete all pending HTLCs> |       |
 |   A   |             ....             |   B   |
 |       |                              |       |
 |       |   <Bob updates his script>   |       |
 |       |                              |       |
 |       |           shutdown(scriptB2) |       |
 |       |<-----------------------------|       |
 |       | shutdown(scriptA2)           |       |
 |       |----------------------------->|       |
 |       |             closing_complete |       |
 |       |          <-------------------|       |
 |       |                              |       |
 |       |  <Alice updates her script>  |       | (*) This is a concurrent update while Bob is sending closing_complete
 |       |                              |       |
 |       | shutdown(scriptA3)           |       |
 |       |------------------->          |       |
 |       |   closing_complete           |       |
 |       |<-------------------          |       | (*) A doesn't answer with closing_sig because B's sig doesn't use scriptA3
 |       |           shutdown(scriptA3) |       |
 |       |          ------------------->|       |
 |       |           shutdown(scriptB2) |       |
 |       |<-----------------------------|       | (*) B answers A's shutdown with his own shutdown, without any changes
 |       |             closing_complete |       |
 |       |<-----------------------------|       | (*) B now uses scriptB2 and scriptA3 for closing_complete 
 |       | closing_sig                  |       |
 |       |----------------------------->|       | (*) A now uses scriptB2 and scriptA3 for closing_sig
 +-------+                              +-------+
 ```

Reply:
Yeah I did realize it's a dual flow - updated the diagram and description, could you verify if it's correct?